### PR TITLE
Update frequency to new form pattern

### PR DESCRIPTION
**Pull request by zecuria**
Updated frequency to the new form pattern

### CODE REVIEW

File: src/components/Form/Form.test.tsx [hunk 1]
```diff
@@ -106,6 +105,11 @@ const ComplexForm = ({
         as={DateInput}
         validate={validateDate}
       />
+      <Field
```
Comment (alifa20) [inline src/components/Form/Form.test.tsx:108]: is there a warning here ?

Comment (zecuria) [reply src/components/Form/Form.test.tsx]: Fixed it

File: src/components/Frequency/Frequency.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+  const fieldErrors = Object.values(error ?? {});
+  const hasErrors = !!error;
+
+  const onFrequencyBlur: React.FocusEventHandler<
+    HTMLInputElement | HTMLSelectElement
+  > = useCallback(
+    (event) => {
+      if (
+        !frequencyRef.current?.contains(event.relatedTarget as HTMLInputElement)
```
Comment (alifa20) [inline src/components/Frequency/Frequency.tsx:90]: Wondering why  we need to cast here? can't we have proper types for event ?

File: src/components/Frequency/Frequency.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    setInternalValue((val) => ({
+      ...val,
+      [key]: updatedValue,
+    }));
+  };
+
+  const frequencyRef = useRef<HTMLDivElement>(null);
+
+  const fieldErrors = Object.values(error ?? {});
+  const hasErrors = !!error;
```
Comment (alifa20) [inline src/components/Frequency/Frequency.tsx:83]: hasError ?

File: src/components/Frequency/Frequency.tsx (hunk 2 above)
Comment (zecuria) [reply src/components/Frequency/Frequency.tsx:90]: It does have proper types for event, however the types for `relatedTarget` are a bit complex and give a number of typescript errors. 

It's just easier to cast in this case for `event.relatedTarget`.

File: src/components/Frequency/index.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+interface StaticProperties {
+  displayName: string;
+  TYPES: string[];
+  VARIANTS: string[];
+  COLORS: string[];
+  DEFAULT_PROPS: typeof FrequencyDeprecated.DEFAULT_PROPS;
+}
+
+// can't be an arrow function due to: https://github.com/yannickcr/eslint-plugin-react/issues/2269
+const Frequency = forwardRef(function InputComponent(
```
Comment (alifa20) [inline src/components/Frequency/index.tsx:24]: how about pulling `function InputComponent` out as FrequencyWithRef 
```
const FrequencyWithRef = (props: FrequencyProps, ref) => ....

const Frequency = forwardRef(FrequencyWithRef)
```

File: src/hooks/useBasisForm/validation/validateFrequency.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import {
+  ALL_FREQUENCY_OPTIONS,
+  FrequencyError,
+  FrequencyValue,
+} from "../../../values";
+import { isInputEmpty, NUMERIC_REGEX } from "./validateInput";
+
+export const validateFrequency = (
+  value: FrequencyValue,
+  props: Record<string, any>
```
Comment (alifa20) [inline src/hooks/useBasisForm/validation/validateFrequency.ts:11]: can't we use `InternalFrequencyProps` instead?

File: src/components/Frequency/index.tsx (hunk 4 above)
Comment (zecuria) [reply src/components/Frequency/index.tsx:24]: I didn't change any of the code from the old frequency. It was like that to begin with

File: src/hooks/useBasisForm/validation/validateInput.ts [hunk 6]
```diff
@@ -3,7 +3,7 @@ import { InternalInputProps } from "../../../components/Input/types";
 export const NUMERIC_REGEX = /^\d*$/;
 export const DECIMAL_REGEX = /^\d*(\.\d{2})?$/;
 
-const isInputEmpty = (value: string) => value.trim() === "";
+export const isInputEmpty = (value: string) => value.trim() === "";
```
Comment (alifa20) [inline src/hooks/useBasisForm/validation/validateInput.ts:6]: `!!value.trim()` can work too . Up to you mate

Comment (zecuria) [reply src/hooks/useBasisForm/validation/validateInput.ts:6]: It was like that from before. I don't mind either way

File: src/utils/test.js [hunk 7]
```diff
@@ -40,13 +40,13 @@ export { userEvent };
 
 // throw if console.error is called in tests
 beforeEach(() => {
-  jest.spyOn(console, "error").mockImplementation(() => {
-    throw new Error();
-  });
+  // jest.spyOn(console, "error").mockImplementation(() => {
```
Comment (alifa20) [inline src/utils/test.js:43]: why

File: src/values.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import { OptionsValues } from "./types";
+
+export interface DateInputValue {
```
Comment (alifa20) [inline src/values.ts:3]: why not moving to `./types` file ?

File: src/components/Frequency/Frequency.tsx [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+  const value = props.value ?? internalValue;
+
+  const onFrequencyChange: React.ChangeEventHandler<
+    HTMLInputElement | HTMLSelectElement
+  > = (event) => {
+    const { name: key, value: updatedValue } = event.target;
+    if (key !== "amount" && key !== "frequency") {
+      return;
+    }
+    onChange?.({
```
Comment (alifa20) [inline src/components/Frequency/Frequency.tsx:69]: nice

File: src/values.ts (hunk 8 above)
Comment (zecuria) [reply src/values.ts:3]: It's easier to export all these out of a single file via `index`, but we probably don't want to export all `types` out of index. So I would have to move them and then re-export them from `index`

Also generally types are being used for more generic multi purpose types, I don't mind doing that but if someone adds additional types they will have to remember to rexport them.

File: src/utils/test.js (hunk 7 above)
Comment (zecuria) [reply src/utils/test.js:43]: my b

File: src/components/Frequency/Frequency.tsx (hunk 9 above)
Comment (strahinjaajvaz) [reply src/components/Frequency/Frequency.tsx:69]: for readability, wouldn't
```javascript
if(onChange) {
    onChange()
}
```
make more sense?

Comment (zecuria) [reply src/components/Frequency/Frequency.tsx:69]: I don't mind either one. I am fine reading the optional param but if you prefer an if statement I am fine with that

File: src/components/Frequency/index.tsx [hunk 10]
```diff
@@ -20,19 +20,19 @@ interface StaticProperties {
   ALL_FREQUENCY_OPTIONS: typeof ALL_FREQUENCY_OPTIONS;
 }
 
-// can't be an arrow function due to: https://github.com/yannickcr/eslint-plugin-react/issues/2269
-const Frequency = forwardRef(function InputComponent(
-  props: FrequencyProps,
-  ref
-) {
+const FrequencyComponent = (props: FrequencyProps, ref: any) => {
```
Comment (alifa20) [inline src/components/Frequency/index.tsx:23]: should be `ref: RefObject<Input>` for the ref type ?

**Review by alifa20 — APPROVED** [review]

**Pull request merged**